pyccel/codegen/printing/cwrappercode.py

```diff
@@ -273,88 +268,187 @@ def get_cast_function_call(self, cast_type, arg):
 
+    # --------------------------------------------------------------------
+    # Functions that take care of collecting (data type, rank) checks and creating the error code
+    # --------------------------------------------------------------------
+    def _generate_TypeError_message(self, variable):
+        """
+        Generate TypeError message from the variable information (datatype, precision)
+        """
+        dtype     = variable.dtype
+
+        if isinstance(dtype, NativeBool):
+            precision = ''
+        if isinstance(dtype, NativeComplex):
+            precision = '{} bit '.format(variable.precision * 2 * 8)
+        else:
+            precision = '{} bit '.format(variable.precision * 8)
+
+        message = '"{var_name} must be {precision}{dtype}"'.format(
+                        var_name  = variable,
+                        precision = precision,
+                        dtype     = self._print(variable.dtype))
+        return message
+
+    def _get_array_type_check(self, variable, collect_var):
+        """
+        Responsible for collecting array type check and creating the
+        corresponding error code
+
+        Parameters:
+        ----------
+        variable     : Variable
+            The optional variable
+        collect_var  : Variable
+            variable which holds the value collected with PyArg_Parsetuple
+
+        Returns:
+        -------
+        check : FunctionCall
+            functionCall responsible for checking the data type
+        error : FunctionCall
+            function call that raise TypeError
+        """
+
+        numpy_dtype = self.find_in_numpy_dtype_registry(variable)
+        arg_dtype   = self.find_in_dtype_registry(self._print(variable.dtype), variable.precision)
+
+        check = PyccelNe(FunctionCall(numpy_get_type, [collect_var]), numpy_dtype)
+        error = PyErr_SetString('PyExc_TypeError', '"{} must be {}"'.format(variable, arg_dtype))
+
+        return check, error
+
+    def _get_scalar_type_check(self, variable, collect_var, error_check = False):
+        """
+        Responsible for collecting numpy and python type check and creating the
+        corresponding error code
+
+        Parameters:
+        ----------
+        variable     : Variable
+            The optional variable
+        collect_var  : Variable
+            variable which holds the value collected with PyArg_Parsetuple
+        error_check  : Bool
+            True if checking the data type and raising error is needed
+
+        Returns:
+        -------
+        python_type_check : FunctionCall
+            functionCall responsible for checking python data type
+            None when error_check is False
+
+        numpy_type_check : FunctionCall or None
+            functionCall responsible for checking numpy data type
+
+        error : FunctionCall or None
+            function call that raise TypeError
+            None when error_check is False
```

Comment:
These are not in the same order as in the returns below

`python_type_check` says it is "None when error_check is False" but it is `LiteralTrue()`

Reply:
done 1039567

Reply:
2b1f9bf